sklearn/model_selection/_search.py

```diff
@@ -513,84 +518,139 @@ def _fit(self, X, y, labels, parameter_iterable):
-                if self.iid:
-                    this_score *= this_n_test_samples
-                    n_test_samples += this_n_test_samples
-                score += this_score
-            if self.iid:
-                score /= float(n_test_samples)
-            else:
-                score /= float(n_splits)
-            scores.append((score, parameters))
-            # TODO: shall we also store the test_fold_sizes?
-            grid_scores.append(_CVScoreTuple(
-                parameters,
-                score,
-                np.array(all_scores)))
-        # Store the computed scores
-        self.grid_scores_ = grid_scores
-
-        # Find the best parameters by comparing on the mean validation score:
-        # note that `sorted` is deterministic in the way it breaks ties
-        best = sorted(grid_scores, key=lambda x: x.mean_validation_score,
-                      reverse=True)[0]
-        self.best_params_ = best.parameters
-        self.best_score_ = best.mean_validation_score
+        n_candidates = len(candidate_params)
+
+        test_scores, test_sample_counts, _ = zip(*out)
+
+        test_scores = np.array(
+            test_scores, dtype=np.float64).reshape(n_candidates, n_splits)
+        # NOTE test_sample counts (weights) remain the same for all candidates
+        test_sample_counts = np.array(test_sample_counts[:n_splits],
+                                      dtype=np.int)
+
+        scorer_name = _get_scorer_name(self.scorer_)
+        results = dict()
+
+        for split_i in range(n_splits):
+            results["%s_test_split_%d" % (scorer_name, split_i)] = (
+                test_scores[:, split_i])
+
+        # Computed the (weighted) mean and std for all the candidates
+        weights = test_sample_counts if self.iid else None
+        means = np.average(test_scores, axis=1, weights=weights)
+
+        stds = np.sqrt(np.average(
+            (test_scores - means[:, np.newaxis]) ** 2, axis=1,
+            weights=weights))
+
+        results["%s_test_mean" % scorer_name] = means
+        results["%s_test_std" % scorer_name] = stds
+
+        ranks = rankdata(means, method='min')
+        # The ranks now need to be reversed (lower rank for higher scores)
+        # (We take max and not n_candidates as there can be ties and same rank
+        # assigned to more than one candidates.)
+        ranks = (max(ranks) + 1) - ranks
+
+        best_index = np.flatnonzero(ranks == 1)[0]
+        best_parameters = candidate_params[best_index]
+        results["%s_test_rank" % scorer_name] = ranks
+
+        for param in parameter_names:
+            # One np.MaskedArray of dtype object per parameter, mask all the
+            # places where the param is not applicable for that candidate.
+            results["param_%s" % param] = np.ma.masked_all(
+                (n_candidates,), dtype=object)
+            for i in range(n_candidates):
+                param_value = candidate_params[i].get(param, None)
```

Comment:
The reason you're using masks is because None is also a valid explicit value for parameters. This line defies the point. What's wrong with good ol' `if param in candidate_params[i]`??